reconcile/openshift_rolebinding.py

```diff
@@ -0,0 +1,226 @@
+    state = AggregatedList()
+
+    for cluster in cluster_store.clusters():
+        api = cluster_store.api(cluster)
+
+        for namespace in cluster_store.namespaces(cluster):
+            for role in cluster_store.namespace_roles(cluster, namespace):
+                rolebindings = api.get_rolebindings(namespace, role)
+
+                members = [
+                    subject[u'name']
+                    for rolebinding in rolebindings
+                    for subject in rolebinding['subjects']
+                    if subject[u'kind'] == u'User'
+                ]
+
+                state.add({
+                    "service": "openshift-rolebinding",
+                    "cluster": cluster,
+                    "namespace": namespace,
+                    "role": role,
+                }, members)
+
+    return state
+
+
+def fetch_desired_state():
+    gqlapi = gql.get_api()
+    result = gqlapi.query(ROLEBINDINGS_QUERY)
+
+    state = AggregatedList()
+
+    def username(m):
+        if m['schema'] == 'access/bot-1.yml':
+            return m.get('github_username_optional')
+        else:
+            return m['github_username']
+
+    for role in result['role']:
+        permissions = filter(
+            lambda p: p.get('service') == 'openshift-rolebinding',
+            role['permissions']
+        )
+
+        if permissions:
+            members = [
+                member for member in
+                (username(m) for m in role['members'])
+                if member is not None
+            ]
+
+            map(lambda p: state.add(p, members), permissions)
+
+    return state
+
+
+class RunnerAction(object):
+    def __init__(self, dry_run, cluster_store):
+        self.dry_run = dry_run
+        self.cluster_store = cluster_store
+
+    def manage_role(self, label, method_name):
+        def action(params, items):
+            cluster = params['cluster']
+            namespace = params['namespace']
+            role = params['role']
+
+            if not self.dry_run:
+                api = self.cluster_store.api(cluster)
+
+            for member in items:
+                logging.info([
+                    label,
+                    member,
+                    cluster,
+                    namespace,
+                    role
+                ])
+
+                if not self.dry_run:
+                    f = getattr(api, method_name)
+                    f(namespace, role, member)
+
+        return action
+
+    def add_role(self):
+        return self.manage_role('add_role', 'add_role_to_user')
+
+    def del_role(self):
+        return self.manage_role('del_role', 'remove_role_from_user')
+
+
+def run(dry_run=False):
+    cluster_store = ClusterStore()
+
+    current_state = fetch_current_state(cluster_store)
+    desired_state = fetch_desired_state()
+
+    # calculate diff
+    diff = current_state.diff(desired_state)
+
+    # ensure all namespace/roles are well known
+    if diff['insert'] != []:
```

Comment:
You should probably compare the length here to zero rather for sanity sake.

Reply:
agreed